README.md
![Screenshot](https://github.com/esmaeil-ahmadipour/esmaeil-ahmadipour/blob/main/upload/packages/flutter_faded_list/banner.png?raw=true "Flutter Faded List Banner")
# Flutter Faded List plugin

This package helps you display a list of your widgets with the blurred image in the first item.

![Screenshot](https://github.com/esmaeil-ahmadipour/esmaeil-ahmadipour/blob/main/upload/packages/flutter_faded_list/preview.gif?raw=true "Flutter Faded List Demo")

### Installation

To use this plugin, add flutter_faded_list in your `pubspec.yaml`

```
dependencies:
  flutter_faded_list: ^0.0.4
```

Or install automatically using this command

```
$ flutter pub add flutter_faded_list
```

### Simple to use

```dart
import 'package:flutter_faded_list/flutter_faded_list.dart';

...
               FadedHorizontalList(
                  blankSpaceWidth: 200,
                  bodyColor: const Color(0xffAD4516),
                  imageWidget: Image.network("https://i.picsum.photos/id/478/536/354.jpg?hmac=adxYyHX8WcCfHkk07quT2s92fbC7vY2QttaeBztwxgI"),
                  children: [
                    for (var i = 0; i < 10; ++i)
                      const Padding(
                      padding: const EdgeInsets.all(70.0),
                      child:Text("  Hello World!  ")),
                  ],
                ),
...
```

### And bonus widget 🎉
Use this widget in FadedHorizontalList for headerWidget property .

```dart
import 'package:flutter_faded_list/flutter_faded_list.dart';

FadedHeaderWidget(title: Text("Sample Header Title")),
...

CHANGELOG.md
## 0.0.1
* initial release

## 0.0.2
* fix some problem on null-safety

## 0.0.3
* add property minOpacityOnImage

## 0.0.4
* Increase performance by removing setState .


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_faded_list_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/flutter_faded_list_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_faded_list/flutter_faded_list.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


example/lib/widgets.dart
import 'package:flutter/material.dart';

class FadedSampleHeaderWidget extends StatelessWidget {
  const FadedSampleHeaderWidget({
    Key? key,
    required this.hotDrink,
    required this.color,
  }) : super(key: key);
  final HotDrink hotDrink;
  final Color color;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(vertical: 8.0, horizontal: 16),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(
                hotDrink == HotDrink.coffee
                    ? "Coffee Category"
                    : "دسته بندی چای",
                style: TextStyle(
                    color: color.computeLuminance() > 0.5
                        ? Colors.black
                        : Colors.white,
                    fontWeight: FontWeight.bold),
              ),
              const Spacer(),
              Text(
                hotDrink == HotDrink.coffee ? "ALL " : " همه",
                style: TextStyle(
                    color: color.computeLuminance() > 0.5
                        ? Colors.black
                        : Colors.white,
                    fontWeight: FontWeight.bold),
              ),
              Icon(Icons.arrow_forward_ios,
                  size: 20,
                  color: color.computeLuminance() > 0.5
                      ? Colors.black
                      : Colors.white),
            ],
          ),
        ),
        const Divider(height: 0),
      ],
    );
  }
}

class FadedSampleItem extends StatelessWidget {
  const FadedSampleItem({
    Key? key,
    required this.hotDrink,
  }) : super(key: key);
  final HotDrink hotDrink;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsetsDirectional.only(
          top: 24, bottom: 24, start: 4.0, end: 4.0),
      child: Container(
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(16.0),
          color: hotDrink == HotDrink.coffee
              ? Colors.brown.shade900
              : Colors.green.shade900,
        ),
        child: hotDrink == HotDrink.coffee
            ? const Icon(Icons.coffee, color: Colors.white, size: 48)
            : const Icon(Icons.coffee_outlined, color: Colors.white, size: 48),
        width: 100,
        height: 100,
      ),
    );
  }
}

class DirectionWidget extends StatelessWidget {
  const DirectionWidget({
    Key? key,
    required this.child,
    required this.textDirection,
  }) : super(key: key);
  final Widget child;
  final TextDirection textDirection;
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: textDirection,
      child: Padding(padding: const EdgeInsets.all(8.0), child: child),
    );
  }
}

enum HotDrink { coffee, tea }

enum LanguageType { rtl, ltr }


lib/flutter_faded_list.dart
library flutter_faded_list;

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

/// [FadedHorizontalList] , this widget used for  display a list of your widgets with the blurred image in the first item.
class FadedHorizontalList extends StatefulWidget {
  const FadedHorizontalList(
      {Key? key,
      required this.children,
      required this.imageWidget,
      this.borderRadius = 16.0,
      this.isDisabledGlow = false,
      this.scrollPhysics = const ScrollPhysics(),
      required this.blankSpaceWidth,
      this.headerWidget,
      required this.bodyColor,
      this.headerColor,
      this.minOpacityOnImage = 0.2})
      : assert(!(blankSpaceWidth < 0),
            '\nYou set the negative value in blankSpaceWidth at FadedHeaderWidget class! ,\nIf you do not need blankSpaceWidth, just set the value to zero.'),
        assert(!(borderRadius < 0),
            '\nYou set the negative value in borderRadius at FadedHeaderWidget class! ,\nIf you do not need borderRadius, just set the value to zero.'),
        assert(!(minOpacityOnImage > 1.0 || minOpacityOnImage < 0.0),
            '\nYou set the minOpacityOnImage value out of range at FadedHeaderWidget class!.\nminOpacityOnImage value must between 0.0 and 1.0 .'),
        super(key: key);

  /// [children] , this widget list used as content and will be scrolled .
  final List<Widget> children;

  /// [imageWidget] , this widget used as background (like image , image network , cachedImage , svg & ETC)
  final Widget imageWidget;

  /// [borderRadius] , this property used for borderRadius of header-widget and main widget (FadedHorizontalList) .
  final double borderRadius;

  /// [bodyColor] , this property used for background of scrollable-widget .
  final Color bodyColor;

  /// [headerColor] , this property used for background of header widget .
  final Color? headerColor;

  /// [isDisabledGlow] , if this property set on true,then hidden glow by scrolling .
  final bool isDisabledGlow;

  /// [scrollPhysics] , this property used for change scroll mode to bounced like iOS or default in android & etc .
  final ScrollPhysics scrollPhysics;

  /// [blankSpaceWidth] , this property used for blank space on first item of content widget list .
  final double blankSpaceWidth;

  /// [headerWidget] , this property used for some time user want to used customized header-widget  .
  final Widget? headerWidget;

  /// [minOpacityOnImage] , this property must set on double number between 0.0 and  1.0 , this property used for minimum opacity on [imageWidget]   .
  final double minOpacityOnImage;

  @override
  State<FadedHorizontalList> createState() => _FadedHorizontalListState();
}

class _FadedHorizontalListState extends State<FadedHorizontalList> {
  late ScrollController _scrollController;
  final GlobalKey _globalKey = GlobalKey();
  ValueNotifier<double> height = ValueNotifier<double>(0.0);
  ValueNotifier<double> scrollPosition = ValueNotifier<double>(0.0);
  @override
  void initState() {
    dynamic dynamicInstance = SchedulerBinding.instance;

    /// cover difference null-safety on SchedulerBinding in flutter3 and older versions.
    dynamicInstance.addPostFrameCallback((timeStamp) {
      /// set height value for using in image_widget.
      height.value = _globalKey.currentContext!.size!.height;
    });

    /// set _scrollController value for using in singleChildScrollView.
    _scrollController = ScrollController();
    _scrollController.addListener(_scrollListener);
    super.initState();
  }

  /// set listener used for detection scroll position.
  _scrollListener() {
    // print("_scrollPosition:$_scrollPosition");
    if (_scrollController.position.pixels <= widget.blankSpaceWidth) {
      scrollPosition.value = _scrollController.position.pixels < 0
          ? 0
          : _scrollController.position.pixels;
    }
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollEndNotification>(
      child: ClipRRect(
        /// this widget handle color and main-size and rounded border  .
        borderRadius: BorderRadius.circular(widget.borderRadius),
        child: Container(
          decoration: BoxDecoration(
            color: widget.bodyColor,
          ),
          key: _globalKey,
          child: Stack(
            // mainAxisSize: MainAxisSize.min,
            children: [
              ValueListenableBuilder<double>(
                  valueListenable: height,
                  builder: (context, double data, child) {
                    return Container(
                      /// this widget , handle image , gradiant on image , image fade  .
                      height: data,
                      foregroundDecoration: BoxDecoration(
                          gradient: LinearGradient(
                              begin: AlignmentDirectional.centerStart,
                              end: AlignmentDirectional.centerEnd,
                              colors: [Colors.transparent, widget.bodyColor])),
                      child: ValueListenableBuilder<double>(
                          valueListenable: scrollPosition,
                          builder: (context, double data, child) {
                            return Opacity(
                              child: widget.imageWidget,
                              opacity: (data < widget.blankSpaceWidth
                                  ? ((widget.blankSpaceWidth - data) /
                                              widget.blankSpaceWidth) <=
                                          widget.minOpacityOnImage
                                      ? widget.minOpacityOnImage
                                      : ((widget.blankSpaceWidth - data) /
                                          widget.blankSpaceWidth)
                                  : widget.minOpacityOnImage),
                            );
                          }),
                    );
                  }),
              Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  if (widget.headerWidget != null)
                    Container(
                      /// this widget , handle header-toolbar , and style around of widget .
                      decoration: BoxDecoration(
                        color: widget.headerColor ??
                            widget.bodyColor.withOpacity(0.5),
                        borderRadius: BorderRadius.only(
                            topLeft: Radius.circular(widget.borderRadius),
                            topRight: Radius.circular(widget.borderRadius)),
                      ),
                      child: widget.headerWidget,
                    ),
                  ScrollConfiguration(
                    /// this widget , handle scroll and scroll-notification , disable scroll-glow , scroll-physics , scroll-direction .
                    behavior: widget.isDisabledGlow
                        ? DisableGlow()
                        : const ScrollBehavior(),
                    child: SingleChildScrollView(
                      physics: widget.scrollPhysics,
                      controller: _scrollController,
                      scrollDirection: Axis.horizontal,
                      child: Padding(
                        padding: EdgeInsetsDirectional.only(
                            start: widget.blankSpaceWidth),
                        child: Row(
                          /// this widget , handle children widgets .
                          children: widget.children,
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// [DisableGlow] ,this widget used for disable glow when user scrolled screen.
class DisableGlow extends ScrollBehavior {
  @override
  Widget buildOverscrollIndicator(
      BuildContext context, Widget child, ScrollableDetails details) {
    return child;
  }
}

/// [FadedHeaderWidget] , this widget is bonus ! you can set header easily on the FadedHorizontalList widgets.
class FadedHeaderWidget extends StatelessWidget {
  const FadedHeaderWidget({
    Key? key,
    this.title,
    this.buttonTitle,
    this.buttonIcon,
    this.onTap,
    this.verticalPadding = 8.0,
    this.horizontalPadding = 16.0,
  })  : assert(!(verticalPadding < 0 || horizontalPadding < 0),
            '\nYou set the negative value in verticalPadding or horizontalPadding at FadedHeaderWidget class! ,\n If you do not need padding, just set the value to zero.'),
        assert(!(title == null && buttonTitle == null && buttonIcon == null),
            '\nall required values in FadedHeaderWidget class is null ,\n at the least mode just set the title value.'),
        super(key: key);

  /// [title] , by this parameter ,  can  set title on the [FadedHeaderWidget] , for example, as a Text("New Category") widget .
  final Widget? title;

  /// [buttonTitle] , by this parameter ,  can  set buttonTitle on the [FadedHeaderWidget] , for example, as a Text("Show All") widget .
  final Widget? buttonTitle;

  /// [buttonIcon] , by this parameter ,  can  set buttonIcon on the [FadedHeaderWidget] , for example, as a Icon(Icons.arrow_forward_outlined) widget .
  final Widget? buttonIcon;

  /// [onTap] , by this parameter ,  can  set Function  onTap on the [FadedHeaderWidget] , for example, when tap on ths [buttonTitle] or [buttonIcon] running this function .
  final Function()? onTap;

  /// [verticalPadding]  , used for padding vertically into [FadedHeaderWidget].
  final double verticalPadding;

  ///  [horizontalPadding] , used for padding horizontally into [FadedHeaderWidget].
  final double horizontalPadding;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: EdgeInsets.symmetric(
              vertical: verticalPadding, horizontal: horizontalPadding),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              if (title != null) title!,
              const Spacer(),
              if (buttonTitle != null || buttonIcon != null)
                GestureDetector(
                  onTap: onTap ?? () {},
                  child: Row(
                    children: [
                      if (buttonTitle != null) buttonTitle!,
                      if (buttonIcon != null) buttonIcon!
                    ],
                  ),
                )
            ],
          ),
        ),
        const Divider(height: 0.0),
      ],
    );
  }
}


example/lib/second_page.dart
import 'package:flutter_faded_list_example/main.dart';
import 'package:flutter_faded_list_example/widgets.dart';
import 'package:flutter/material.dart';
import 'package:flutter_faded_list/flutter_faded_list.dart';

class SecondPage extends StatelessWidget {
  const SecondPage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.format_textdirection_r_to_l),
        onPressed: () {
          Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => FirstPage(title: title)),
          );
        },
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            const Text("TextDirection  RTL"),
            DirectionWidget(
              textDirection: TextDirection.rtl,
              child: FadedHorizontalList(
                blankSpaceWidth: 200,
                bodyColor: const Color(0xff044b1f),
                imageWidget: Image.network(
                    "https://images.unsplash.com/photo-1564890369478-c89ca6d9cde9?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170"),
                children: [
                  for (var i = 0; i < 10; ++i)
                    const FadedSampleItem(hotDrink: HotDrink.tea)
                ],
              ),
            ),
            DirectionWidget(
                textDirection: TextDirection.rtl,
                child: FadedHorizontalList(
                  headerColor: const Color(0xff0c4e23),
                  blankSpaceWidth: 200,
                  bodyColor: const Color(0xff00ff12),
                  borderRadius: 20.0,
                  imageWidget: Image.network(
                      "https://images.unsplash.com/photo-1564890369478-c89ca6d9cde9?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170"),
                  headerWidget: const FadedSampleHeaderWidget(
                      hotDrink: HotDrink.tea, color: Color(0xff0c4e23)),
                  children: [
                    for (var i = 0; i < 10; ++i)
                      const FadedSampleItem(hotDrink: HotDrink.tea)
                  ],
                )),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter_faded_list_example/second_page.dart';
import 'package:flutter_faded_list_example/widgets.dart';
import 'package:flutter/material.dart';
import 'package:flutter_faded_list/flutter_faded_list.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Flutter Faded List',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const FirstPage(
          title: "Flutter Faded List",
        ));
  }
}

class FirstPage extends StatelessWidget {
  const FirstPage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.format_textdirection_l_to_r),
        onPressed: () {
          Navigator.pushReplacement(
            context,
            MaterialPageRoute(builder: (context) => SecondPage(title: title)),
          );
        },
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            const Text("TextDirection  LTR"),
            DirectionWidget(
              textDirection: TextDirection.ltr,
              child: FadedHorizontalList(
                blankSpaceWidth: 200,
                bodyColor: const Color(0xffAD4516),
                imageWidget: Image.network(
                    "https://images.unsplash.com/photo-1561336313-0bd5e0b27ec8?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170"),
                children: [
                  for (var i = 0; i < 10; ++i)
                    const FadedSampleItem(hotDrink: HotDrink.coffee)
                ],
              ),
            ),
            DirectionWidget(
              textDirection: TextDirection.ltr,
              child: FadedHorizontalList(
                headerColor: const Color(0xffAD4516),
                blankSpaceWidth: 200,
                bodyColor: const Color(0xffd27900),
                borderRadius: 20.0,
                imageWidget: Image.network(
                    "https://images.unsplash.com/photo-1561336313-0bd5e0b27ec8?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170"),
                headerWidget: const FadedSampleHeaderWidget(
                    hotDrink: HotDrink.coffee, color: Color(0xffAD4516)),
                children: [
                  for (var i = 0; i < 10; ++i)
                    const FadedSampleItem(hotDrink: HotDrink.coffee)
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
